internal/docs/design.md

```diff
@@ -374,6 +375,106 @@ b, err := blob.NewBucket(d, blob.FeatureUnicodeNames)
 Design discussions regarding enforcing portability are ongoing; we welcome input
 on the [mailing list](https://groups.google.com/forum/#!forum/go-cloud).
 
+### Strings
+
+Providers often differ on what they accept in particular strings (e.g., blob
+names, metadata keys, etc.). A couple of specific examples:
+
+*   Azure Blob only
+    [accepts C# identifiers](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-properties-metadata)
+    as metadata keys.
+*   S3 drops double slashes in blob names (e.g., `foo//bar` will end up being
+    saved as `foo/bar`).
+
+These differences lead to a loss of portability and predictability for users.
+
+We considered restricting Go CDK's APIs to strings that all providers support.
+For example, we could assert that Go CDK's blob only supports ASCII plus `/` for
+blob names (and no `//`!). However, such a rule would mean that we couldn't
+handle existing strings that were created through some mechanism other than
+through Go CDK APIs. For example, an existing blob in S3 with a unicode name
+wouldn't be visible in Go CDK. Or maybe it would be visible, but you wouldn't be
+able to write to it. For metadata keys, rules like this might mean that you
+could read metadata attributes via the Go CDK and then write them back, and
+either lose data (if previously-existing invalid keys were dropped) or get an
+error. Overall, we decided that this approach is unacceptable.
+
+Instead, we insist that Go CDK can handle any string, and force provider
```

Comment:
Do we have some less ambiguous terms to distinguish between "provider implementations" and "underlying provider"?

Reply:
I think "driver" instead of "provider implementation" is more consistent with the rest of the doc; fixed.